keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1181 @@
+> - Is there another option?
+> This has consistency and performance implications.
+>
+> <<[/UNRESOLVED]>>
+
+We will not support “adoption” of existing applysets initially, other than by applying “over the top”.  Based on user feedback, we may require a flag to adopt existing objects.  Note that adoption is not trivial, in that different users may expect different behaviors with regard to the GKs selected or the treatment of objects having/lacking the `last-application-configuration` annotation.
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> Expand on adoption (taking over management of a set created by another tool)
+> vs migration (taking over management of a set created by the old pruning mechanism)
+> and how kubectl will or won't facilitate these flows.
+>
+> <<[/UNRESOLVED]>>
+
+### Tooling Interoperability
+
+There is a rich ecosystem of existing tooling that we hope will adopt these labels and annotations.  So that different tooling can interoperate smoothly, we define some requirements for safe interoperability here.
+
+For read operations, we expect that using different tooling shall generally be safe.  As these labels do not collide with existing tooling, we would expect that objects installed with existing tooling would be invisible to the porcelain tooling until they had been updated to include the labels.  We do not propose to implement “bridges” to existing tooling, rather as the proposal here is lightweight and small, it makes more sense to update the existing tooling.  We may add warnings such as “applysets using an old version of X detected, upgrade to v123 of X to work with those applysets”.
+
+For write operations, we need to be more careful.  Deleting an applyset using the “wrong tool” should be safe, but we will likely include a confirmation if deleting an applyset using the “wrong tool”, particularly unknown tools.  We expect that porcelain tools may define richer behavior on delete, so this is the equivalent of pulling the power cable on an applyset instead of performing a clean shutdown.
+
+We do not believe that update operations are safe if using the “wrong tool”, because that tooling may have additional metadata that would then not be updated.  Tooling should generally reject applying on top of unknown applysets.  PorcelainTooling may choose to recognize other tooling and implement specific logic there; in particular this may be useful for moving between different major versions of the same tooling.  We may implement a `--force` flag, but this would likely be logically equivalent in outcome to a full applyset deletion and recreation, though with the potential (but not the guarantee) to be less disruptive.
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+>  How should we recognize that we’re using the “wrong’ tool?
+> Should we have something like `applyset.k8s.io/tooling: helm/v2.0.6` on the applyset?
+>
+> <<[/UNRESOLVED]>>
```

Comment:
Since you're introducing a meta object holding information about applyset contents, have you considered versioning the metadata?